File: windows/VpsQuota/Storage/SqliteStore.cs
namespace VpsQuota.Storage;

using System.IO;
using Microsoft.Data.Sqlite;
using VpsQuota.Models;

/// <summary>
/// 日流量与采集日志的本地持久化。
///
/// 存在的理由：上游数据窗口有限（Vultr API 约 30 天、vnstat 日表默认保留 62 天），
/// 每次采集都覆盖式 upsert 到本地后，历史可以无限累积，单次采集失败也不会让界面变空。
///
/// 所有访问都串行化（<see cref="_gate"/>），因为刷新时多台服务器是并发采集的，
/// 而同一个 SQLite 连接不能被多个线程同时使用。
/// </summary>
public sealed class SqliteStore : IAsyncDisposable
{
    private readonly SqliteConnection _connection;
    private readonly SemaphoreSlim _gate = new(1, 1);

    public SqliteStore(string path)
    {
        var directory = Path.GetDirectoryName(path);
        if (!string.IsNullOrEmpty(directory)) Directory.CreateDirectory(directory);

        _connection = new SqliteConnection(new SqliteConnectionStringBuilder
        {
            DataSource = path,
            Mode = SqliteOpenMode.ReadWriteCreate,
        }.ToString());
        _connection.Open();
        Migrate();
    }

    /// <summary>建表语句与 shared/schema.sql 保持一致，改动时两处都要改。</summary>
    private void Migrate()
    {
        using var command = _connection.CreateCommand();
        command.CommandText = """
            PRAGMA journal_mode = WAL;
            -- 诊断用的 CLI 与本应用可能同时开着同一个库（README 就是这么建议排查的）。
            -- 没有 busy handler 时，SQLite 撞上锁会立刻返回 SQLITE_BUSY，
            -- 一次本来成功的采集就被记成失败、数据也丢了。等 5 秒足够让对方写完。
            PRAGMA busy_timeout = 5000;
            CREATE TABLE IF NOT EXISTS daily_usage (
                server_id TEXT    NOT NULL,
                day       TEXT    NOT NULL,
                rx_bytes  INTEGER NOT NULL,
                tx_bytes  INTEGER NOT NULL,
                PRIMARY KEY (server_id, day)
            );
            CREATE INDEX IF NOT EXISTS idx_daily_usage_day ON daily_usage (day);
            CREATE TABLE IF NOT EXISTS fetch_log (
                server_id  TEXT    NOT NULL,
                fetched_at INTEGER NOT NULL,
                ok         INTEGER NOT NULL,
                error      TEXT
            );
            CREATE INDEX IF NOT EXISTS idx_fetch_log_server ON fetch_log (server_id, fetched_at DESC);
            CREATE TABLE IF NOT EXISTS server_meta (
                server_id         TEXT NOT NULL,
                reported_quota_gb REAL,
                PRIMARY KEY (server_id)
            );
            """;
        command.ExecuteNonQuery();
    }

    // MARK: 写入

    /// <summary>
    /// 覆盖式写入若干天的流量。
    ///
    /// 必须是覆盖（DO UPDATE）而非忽略：当天的数据在一天之内会被反复刷新并持续增长。
    /// </summary>
    public async Task UpsertAsync(string serverId, IReadOnlyList<DailyUsage> days)
    {
        if (days.Count == 0) return;

        await _gate.WaitAsync().ConfigureAwait(false);
        try
        {
            using var transaction = _connection.BeginTransaction();
            using var command = _connection.CreateCommand();
            command.Transaction = transaction;
            command.CommandText = """
                INSERT INTO daily_usage (server_id, day, rx_bytes, tx_bytes)
                VALUES ($server, $day, $rx, $tx)
                ON CONFLICT(server_id, day) DO UPDATE SET
                    rx_bytes = excluded.rx_bytes,
                    tx_bytes = excluded.tx_bytes
                """;
            var pServer = command.Parameters.Add("$server", Microsoft.Data.Sqlite.SqliteType.Text);
            var pDay = command.Parameters.Add("$day", Microsoft.Data.Sqlite.SqliteType.Text);
            var pRx = command.Parameters.Add("$rx", Microsoft.Data.Sqlite.SqliteType.Integer);
            var pTx = command.Parameters.Add("$tx", Microsoft.Data.Sqlite.SqliteType.Integer);

            pServer.Value = serverId;
            foreach (var d in days)
            {
                pDay.Value = d.Day;
                pRx.Value = d.RxBytes;
                pTx.Value = d.TxBytes;
                command.ExecuteNonQuery();
            }
            transaction.Commit();
        }
        finally
        {
            _gate.Release();
        }
    }

    public async Task LogFetchAsync(string serverId, DateTime at, bool ok, string? error)
    {
        await _gate.WaitAsync().ConfigureAwait(false);
        try
        {
            using var command = _connection.CreateCommand();
            command.CommandText =
                "INSERT INTO fetch_log (server_id, fetched_at, ok, error) VALUES ($s, $t, $ok, $e)";
            command.Parameters.AddWithValue("$s", serverId);
            // 必须先定 Kind：DateTimeOffset(DateTime) 对 Unspecified 按本地时区解释，
            // 时间戳会整体偏移数小时，「上次刷新于」就跟着错。
            command.Parameters.AddWithValue("$t", ToUnixSeconds(at));
            command.Parameters.AddWithValue("$ok", ok ? 1 : 0);
            command.Parameters.AddWithValue("$e", (object?)error ?? DBNull.Value);
            command.ExecuteNonQuery();
        }
        finally
        {
            _gate.Release();
        }
    }

    // MARK: 读取

    /// <summary>
    /// 读取 <c>[from, toExclusive)</c> 区间的日流量，按日期升序。
    /// 参数是 yyyy-MM-dd 字符串 —— 该格式字典序即时间序，所以直接用文本比较即可。
    /// </summary>
    public async Task<List<DailyUsage>> DaysAsync(string serverId, string from, string toExclusive)
    {
        await _gate.WaitAsync().ConfigureAwait(false);
        try
        {
            using var command = _connection.CreateCommand();
            command.CommandText = """
                SELECT day, rx_bytes, tx_bytes FROM daily_usage
                WHERE server_id = $s AND day >= $from AND day < $to
                ORDER BY day ASC
                """;
            command.Parameters.AddWithValue("$s", serverId);
            command.Parameters.AddWithValue("$from", from);
            command.Parameters.AddWithValue("$to", toExclusive);

            var result = new List<DailyUsage>();
            using var reader = command.ExecuteReader();
            while (reader.Read())
            {
                result.Add(new DailyUsage(reader.GetString(0), reader.GetInt64(1), reader.GetInt64(2)));
            }
            return result;
        }
        finally
        {
            _gate.Release();
        }
    }

    /// <summary>最近一次成功采集的时间。从未成功过则返回 null。</summary>
    public async Task<DateTime?> LastSuccessAsync(string serverId)
    {
        await _gate.WaitAsync().ConfigureAwait(false);
        try
        {
            using var command = _connection.CreateCommand();
            command.CommandText = "SELECT MAX(fetched_at) FROM fetch_log WHERE server_id = $s AND ok = 1";
            command.Parameters.AddWithValue("$s", serverId);

            var value = command.ExecuteScalar();
            if (value is null || value == DBNull.Value) return null;
            return DateTimeOffset.FromUnixTimeSeconds(Convert.ToInt64(value)).UtcDateTime;
        }
        finally
        {
            _gate.Release();
        }
    }

    /// <summary>
    /// 记下上游报告的配额（目前只有 Vultr 有）。传 null 表示"这次没拿到"，不覆盖已有值。
    /// </summary>
    public async Task SetReportedQuotaAsync(string serverId, double? quotaGB)
    {
        if (quotaGB is not { } value) return;

        await _gate.WaitAsync().ConfigureAwait(false);
        try
        {
            using var command = _connection.CreateCommand();
            command.CommandText =
                "INSERT INTO server_meta (server_id, reported_quota_gb) VALUES ($s, $q) " +
                "ON CONFLICT(server_id) DO UPDATE SET reported_quota_gb = excluded.reported_quota_gb";
            command.Parameters.AddWithValue("$s", serverId);
            command.Parameters.AddWithValue("$q", value);
            command.ExecuteNonQuery();
        }
        finally
        {
            _gate.Release();
        }
    }

    /// <summary>读回上游报告的配额。没记录过则为 null。</summary>
    public async Task<double?> ReportedQuotaAsync(string serverId)
    {
        await _gate.WaitAsync().ConfigureAwait(false);
        try
        {
            using var command = _connection.CreateCommand();
            command.CommandText = "SELECT reported_quota_gb FROM server_meta WHERE server_id = $s";
            command.Parameters.AddWithValue("$s", serverId);

            var value = command.ExecuteScalar();
            if (value is null || value == DBNull.Value) return null;
            return Convert.ToDouble(value);
        }
        finally
        {
            _gate.Release();
        }
    }

    /// <summary>最近一次采集的错误。若最近一次是成功的则返回 null。</summary>
    /// <remarks>
    /// 语义与内存里那份 LastErrors 一致：成功一次就把错误清掉。
    /// 有了它，应用重启后不必等第一轮采集跑完，也能如实显示上次的失败原因（P5）。
    /// </remarks>
    public async Task<string?> LastErrorAsync(string serverId)
    {
        await _gate.WaitAsync().ConfigureAwait(false);
        try
        {
            using var command = _connection.CreateCommand();
            command.CommandText =
                "SELECT ok, error FROM fetch_log WHERE server_id = $s " +
                "ORDER BY fetched_at DESC LIMIT 1";
            command.Parameters.AddWithValue("$s", serverId);

            using var reader = command.ExecuteReader();
            if (!reader.Read()) return null;
            if (reader.GetInt32(0) != 0) return null;      // 最近一次是成功的
            return reader.IsDBNull(1) ? null : reader.GetString(1);
        }
        finally
        {
            _gate.Release();
        }
    }

    /// <summary>
    /// 折算成 Unix 秒。Local 按时区换算；Unspecified 视作已经是 UTC ——
    /// 应用内的时间全部来自 DateTime.UtcNow，但类型系统拦不住误传，这里兜一层。
    /// </summary>
    private static long ToUnixSeconds(DateTime value) => value.Kind switch
    {
        DateTimeKind.Utc => new DateTimeOffset(value).ToUnixTimeSeconds(),
        DateTimeKind.Local => new DateTimeOffset(value).ToUniversalTime().ToUnixTimeSeconds(),
        _ => new DateTimeOffset(DateTime.SpecifyKind(value, DateTimeKind.Utc)).ToUnixTimeSeconds(),
    };

    /// <summary>清理过期的采集日志，避免文件无限增长。默认保留 90 天。</summary>
    public async Task PruneFetchLogAsync(int olderThanDays, DateTime now)
    {
        await _gate.WaitAsync().ConfigureAwait(false);
        try
        {
            using var command = _connection.CreateCommand();
            command.CommandText = "DELETE FROM fetch_log WHERE fetched_at < $cutoff";
            command.Parameters.AddWithValue("$cutoff", ToUnixSeconds(now.AddDays(-olderThanDays)));
            command.ExecuteNonQuery();
        }
        finally
        {
            _gate.Release();
        }
    }

    public async ValueTask DisposeAsync()
    {
        await _connection.DisposeAsync().ConfigureAwait(false);
        _gate.Dispose();
    }
}

File: macos/Sources/VPSQuotaCore/Storage/SQLiteStore.swift
import Foundation
import SQLite3

/// SQLite 需要的字符串绑定语义：告诉 SQLite 自行拷贝入参，
/// 否则 Swift 侧的临时字符串在 step 之前就可能被释放。
private let SQLITE_TRANSIENT = unsafeBitCast(-1, to: sqlite3_destructor_type.self)

public enum StoreError: LocalizedError {
    case open(String)
    case sql(String)

    public var errorDescription: String? {
        switch self {
        case .open(let m): return "无法打开本地数据库：\(m)"
        case .sql(let m): return "数据库操作失败：\(m)"
        }
    }
}

/// 日流量与采集日志的本地持久化。
///
/// 存在的理由：上游数据窗口有限（Vultr API 约 30 天、vnstat 日表默认保留 62 天），
/// 每次采集都覆盖式 upsert 到本地后，历史可以无限累积，单次采集失败也不会让界面变空。
///
/// 用 `actor` 串行化所有访问，因为 SQLite 连接不是线程安全的，
/// 而刷新时四台服务器是并发采集的。
public actor SQLiteStore {
    /// 连接句柄本身在 init 之后不再变动，所以标 `nonisolated(unsafe)` 让 deinit 能关闭它；
    /// 对该连接的所有**使用**仍然由 actor 串行化 —— SQLite 连接不是线程安全的。
    private nonisolated(unsafe) let db: OpaquePointer

    public init(path: URL) throws {
        try FileManager.default.createDirectory(
            at: path.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )
        var handle: OpaquePointer?
        guard sqlite3_open(path.path, &handle) == SQLITE_OK, let handle else {
            let msg = handle.map { String(cString: sqlite3_errmsg($0)) } ?? "unknown"
            sqlite3_close(handle)
            throw StoreError.open(msg)
        }
        self.db = handle

        // 诊断用的 CLI 与图形界面可能同时开着同一个库（README 就是这么建议排查的）。
        // 没有 busy handler 时，SQLite 撞上锁会立刻返回 SQLITE_BUSY，
        // upsert 的 BEGIN IMMEDIATE 直接失败 —— 一次本来成功的采集被记成失败、数据也丢了。
        sqlite3_busy_timeout(handle, 5_000)

        try Self.migrate(handle)
    }

    deinit {
        sqlite3_close(db)
    }

    /// 建表语句与 shared/schema.sql 保持一致，改动时两处都要改。
    ///
    /// 是静态函数而非实例方法：actor 的 `init` 处于 nonisolated 上下文，无法调用隔离成员。
    private static func migrate(_ handle: OpaquePointer) throws {
        let sql = """
        PRAGMA journal_mode = WAL;
        CREATE TABLE IF NOT EXISTS daily_usage (
            server_id TEXT    NOT NULL,
            day       TEXT    NOT NULL,
            rx_bytes  INTEGER NOT NULL,
            tx_bytes  INTEGER NOT NULL,
            PRIMARY KEY (server_id, day)
        );
        CREATE INDEX IF NOT EXISTS idx_daily_usage_day ON daily_usage (day);
        CREATE TABLE IF NOT EXISTS fetch_log (
            server_id  TEXT    NOT NULL,
            fetched_at INTEGER NOT NULL,
            ok         INTEGER NOT NULL,
            error      TEXT
        );
        CREATE INDEX IF NOT EXISTS idx_fetch_log_server ON fetch_log (server_id, fetched_at DESC);
        """
        try exec(handle, sql)
    }

    private static func exec(_ handle: OpaquePointer?, _ sql: String) throws {
        var errMsg: UnsafeMutablePointer<CChar>?
        guard sqlite3_exec(handle, sql, nil, nil, &errMsg) == SQLITE_OK else {
            let msg = errMsg.map { String(cString: $0) } ?? "unknown"
            sqlite3_free(errMsg)
            throw StoreError.sql(msg)
        }
    }

    private func exec(_ sql: String) throws {
        try Self.exec(db, sql)
    }

    private func lastErrorMessage() -> String {
        String(cString: sqlite3_errmsg(db))
    }

    // MARK: - 写入

    /// 覆盖式写入若干天的流量。
    ///
    /// 必须是覆盖（`DO UPDATE`）而非忽略：当天的数据在一天之内会被反复刷新并持续增长。
    public func upsert(serverId: String, days: [DailyUsage]) throws {
        guard !days.isEmpty else { return }

        try exec("BEGIN IMMEDIATE")
        do {
            let sql = """
            INSERT INTO daily_usage (server_id, day, rx_bytes, tx_bytes)
            VALUES (?, ?, ?, ?)
            ON CONFLICT(server_id, day) DO UPDATE SET
                rx_bytes = excluded.rx_bytes,
                tx_bytes = excluded.tx_bytes
            """
            var stmt: OpaquePointer?
            guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
                throw StoreError.sql(lastErrorMessage())
            }
            defer { sqlite3_finalize(stmt) }

            for d in days {
                sqlite3_reset(stmt)
                sqlite3_clear_bindings(stmt)
                sqlite3_bind_text(stmt, 1, serverId, -1, SQLITE_TRANSIENT)
                sqlite3_bind_text(stmt, 2, d.day, -1, SQLITE_TRANSIENT)
                sqlite3_bind_int64(stmt, 3, d.rxBytes)
                sqlite3_bind_int64(stmt, 4, d.txBytes)
                guard sqlite3_step(stmt) == SQLITE_DONE else {
                    throw StoreError.sql(lastErrorMessage())
                }
            }
            try exec("COMMIT")
        } catch {
            try? exec("ROLLBACK")
            throw error
        }
    }

    public func logFetch(serverId: String, at date: Date, ok: Bool, error: String?) throws {
        let sql = "INSERT INTO fetch_log (server_id, fetched_at, ok, error) VALUES (?, ?, ?, ?)"
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
            throw StoreError.sql(lastErrorMessage())
        }
        defer { sqlite3_finalize(stmt) }

        sqlite3_bind_text(stmt, 1, serverId, -1, SQLITE_TRANSIENT)
        sqlite3_bind_int64(stmt, 2, Int64(date.timeIntervalSince1970))
        sqlite3_bind_int(stmt, 3, ok ? 1 : 0)
        if let error {
            sqlite3_bind_text(stmt, 4, error, -1, SQLITE_TRANSIENT)
        } else {
            sqlite3_bind_null(stmt, 4)
        }
        guard sqlite3_step(stmt) == SQLITE_DONE else {
            throw StoreError.sql(lastErrorMessage())
        }
    }

    // MARK: - 读取

    /// 读取 `[from, toExclusive)` 区间的日流量，按日期升序。
    ///
    /// 参数是 `YYYY-MM-DD` 字符串 —— 该格式字典序即时间序，所以直接用文本比较即可。
    public func days(serverId: String, from: String, toExclusive: String) throws -> [DailyUsage] {
        let sql = """
        SELECT day, rx_bytes, tx_bytes FROM daily_usage
        WHERE server_id = ? AND day >= ? AND day < ?
        ORDER BY day ASC
        """
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
            throw StoreError.sql(lastErrorMessage())
        }
        defer { sqlite3_finalize(stmt) }

        sqlite3_bind_text(stmt, 1, serverId, -1, SQLITE_TRANSIENT)
        sqlite3_bind_text(stmt, 2, from, -1, SQLITE_TRANSIENT)
        sqlite3_bind_text(stmt, 3, toExclusive, -1, SQLITE_TRANSIENT)

        var result: [DailyUsage] = []
        while sqlite3_step(stmt) == SQLITE_ROW {
            guard let dayPtr = sqlite3_column_text(stmt, 0) else { continue }
            result.append(DailyUsage(
                day: String(cString: dayPtr),
                rxBytes: sqlite3_column_int64(stmt, 1),
                txBytes: sqlite3_column_int64(stmt, 2)
            ))
        }
        return result
    }

    /// 最近一次成功采集的时间。从未成功过则返回 nil。
    public func lastSuccess(serverId: String) throws -> Date? {
        let sql = "SELECT MAX(fetched_at) FROM fetch_log WHERE server_id = ? AND ok = 1"
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
            throw StoreError.sql(lastErrorMessage())
        }
        defer { sqlite3_finalize(stmt) }

        sqlite3_bind_text(stmt, 1, serverId, -1, SQLITE_TRANSIENT)
        guard sqlite3_step(stmt) == SQLITE_ROW,
              sqlite3_column_type(stmt, 0) != SQLITE_NULL else { return nil }
        return Date(timeIntervalSince1970: TimeInterval(sqlite3_column_int64(stmt, 0)))
    }

    /// 清理过期的采集日志，避免文件无限增长。默认保留 90 天。
    public func pruneFetchLog(olderThan days: Int = 90, now: Date = Date()) throws {
        let cutoff = Int64(now.addingTimeInterval(-Double(days) * 86_400).timeIntervalSince1970)
        try exec("DELETE FROM fetch_log WHERE fetched_at < \(cutoff)")
    }
}
